types/src/signed.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{Index, NodeCount, NodeIndex, NodeMap};
+use async_trait::async_trait;
+use codec::Codec;
+use std::fmt::Debug;
+
+/// The type used as a signature.
+///
+/// The Signature typically does not contain the index of the node who signed the data.
+pub trait Signature: Debug + Clone + Codec + Send + Sync + Eq + 'static {}
+
+impl<T: Debug + Clone + Codec + Send + Sync + Eq + 'static> Signature for T {}
+
+/// Abstraction of the signing data and verifying signatures.
+///
+/// A typical implementation of KeyBox would be a collection of `N` public keys,
+/// an index `i` and a single private key corresponding to the public key number `i`.
+/// The meaning of sign is then to produce a signature `s` using the given private key,
+/// and `verify(msg, s, j)` is to verify whether the signature s under the message msg is
+/// correct with respect to the public key of the jth node.
+#[async_trait]
+pub trait KeyBox: Index + Clone + Send + Sync + 'static {
+    type Signature: Signature;
+
+    /// Returns the total number of known public keys.
+    fn node_count(&self) -> NodeCount;
+    /// Signs a message `msg`.
+    async fn sign(&self, msg: &[u8]) -> Self::Signature;
+    /// Verifies whether a node with `index` correctly signed the message `msg`.
+    fn verify(&self, msg: &[u8], sgn: &Self::Signature, index: NodeIndex) -> bool;
+}
+
+/// A type to which signatures can be aggregated.
+///
+/// Any signature can be added to multisignature.
+/// After adding sufficiently many signatures, the partial multisignature becomes a "complete"
+/// multisignature.
+/// Whether a multisignature is complete, can be verified with [`MultiKeychain::is_complete`] method.
+/// The signature and the index passed to the `add_signature` method are required to be valid.
+pub trait PartialMultisignature: Signature {
+    type Signature: Signature;
+    /// Adds the signature.
+    #[must_use]
```

Comment:
Do you have something specific in mind? I'm not sure if we should even have `#[must_use]` here...

Reply:
Since we already have them here I would say adding a message to them is a good idea. And they definitely should be here.

Reply:
Sorry, didn't realize this was waiting for me! As discussed in today's standup, I think #[must_use] is applicable in many situations and should generally come with a specific message. However, if you think it would be better to remove it in this case, I'm not strongly opposed.

Reply:
As @timorl accepted the PR in its current state, I'll mark this as resolved and address the issue later.